src/coreclr/src/pal/src/map/virtual.cpp

```diff
@@ -1760,6 +1760,23 @@ VirtualProtect(
     return bRetVal;
 }
 
+#if defined(TARGET_OSX) && defined(TARGET_ARM64)
+bool
+PAL_JITWriteEnableHolder::JITWriteEnable(bool writeEnable)
+{
+    // Use a thread local to track per thread JIT Write enable state
+    // Threads start with MAP_JIT pages readable and executable (R-X) by default.
+    thread_local int enabled = 0xdeadbeef;
```

Comment:
@janvorli 
This was originally 
```c++
thread_local bool enabled = false;
```

This effectively made the assumption that new threads started with JIT write disabled and execute enabled.  While this assumption seems logical to me, it doesn't match the minimal documentation in the `pthread_jit_write_protect_np` man page.

```
pthread_jit_write_protect_np() must be called to toggle per-thread write protection on the MAP_JIT region 
before the thread writes to or executes from the MAP_JIT region.
```

It seemed to work with my original assumption, but I changed it during debugging and decided to keep it.

Reply:
I understand this sentence as meaning that you need to call it before writing and after writing. But anyways, we always have to first write some code to the memory to execute it, so whether the sentence should be understood your way or my way doesn't really matter.

Reply:
Except when we spawn a new thread and code is already written...

Reply:
So you think that for every new thread, this function needs to be called to make code jitted on other threads executable? That would sound like a weird design of the mechanism behind pthread_jit_write_protect_np to me, but I may be wrong, of course.

Reply:
I agree it sounds weird...  The docs are unclear...  We could look at the BSD kernel source to see the default...  My guess is it would be one or the other rather than random...

Reply:
> this function needs to be called to make code jitted on other threads executable

This was confirmed in testing.

Reply:
If it is the case, it means that we have a problem. There are common situations where the JITed code is called on a new threads and we do not have a good place to call this function on the thread before JITed code starts executing.

Reply:
What are such cases? I have thought that the only cases were the ones when a managed callback was called from an external native thread and those we can handle.

Reply:
- Callbacks created by `Marshal.GetFunctionPointerForDelegate`. They point to `UMEntryThunk` that is `MAP_JIT` code.  
- `UnmanagedCallersOnly` function pointers: They point to either precode or actual JITed code. It is `MAP_JIT` code in both cases.

Reply:
> My guess is it would be one or the other rather than random

Maybe the value is inherited from the parent thread?

Reply:
> it means that we have a problem

We don't seem to have any coverage for the problematic cases.  At least not in p1 tests...

Reply:
We seem to need to setup a managed thread object for any native threads.  I put the initialization in `SetupThread()`.  So far that seems to work for all the existing p1 tests.

I'll reopen #41991.